File: Game-1-modular/data/models/recipes.py
"""Recipe and placement data models"""

from dataclasses import dataclass, field
from typing import List, Dict


@dataclass
class Recipe:
    """Crafting recipe definition"""
    recipe_id: str
    output_id: str
    output_qty: int
    station_type: str
    station_tier: int
    inputs: List[Dict]
    grid_size: str = "3x3"
    mini_game_type: str = ""
    metadata: Dict = field(default_factory=dict)
    # Enchanting-specific fields
    is_enchantment: bool = False
    enchantment_name: str = ""
    applicable_to: List[str] = field(default_factory=list)
    effect: Dict = field(default_factory=dict)


@dataclass
class PlacementData:
    """Universal placement data structure for all crafting disciplines"""
    recipe_id: str
    discipline: str  # smithing, alchemy, refining, engineering, adornments

    # Smithing & Enchanting: Grid-based placement
    grid_size: str = ""  # "3x3", "5x5", "12x12", etc.
    placement_map: Dict[str, str] = field(default_factory=dict)  # "x,y" -> materialId

    # Refining: Hub-and-spoke
    core_inputs: List[Dict] = field(default_factory=list)  # Center slots
    surrounding_inputs: List[Dict] = field(default_factory=list)  # Surrounding modifiers

    # Alchemy: Sequential
    ingredients: List[Dict] = field(default_factory=list)  # [{slot, materialId, quantity}]

    # Engineering: Slot types
    slots: List[Dict] = field(default_factory=list)  # [{type, materialId, quantity}]

    # Enchanting: Pattern-based
    pattern: List[Dict] = field(default_factory=list)  # Pattern vertices/shapes

    # Metadata
    narrative: str = ""
    output_id: str = ""
    station_tier: int = 1

File: Unity/Assets/Scripts/Game1.Data/Models/Recipes.cs
// Game1.Data.Models.Recipes
// Migrated from: data/models/recipes.py (52 lines)
// Phase: 1 - Foundation
// Contains: Recipe, PlacementData

using System;
using System.Collections.Generic;
using Newtonsoft.Json;

namespace Game1.Data.Models
{
    /// <summary>
    /// Crafting recipe definition. Pure data - no methods.
    /// </summary>
    [Serializable]
    public class Recipe
    {
        [JsonProperty("recipeId")]
        public string RecipeId { get; set; }

        [JsonProperty("outputId")]
        public string OutputId { get; set; }

        [JsonProperty("outputQty")]
        public int OutputQty { get; set; }

        [JsonProperty("stationType")]
        public string StationType { get; set; }

        [JsonProperty("stationTier")]
        public int StationTier { get; set; }

        [JsonProperty("inputs")]
        public List<Dictionary<string, object>> Inputs { get; set; } = new();

        [JsonProperty("gridSize")]
        public string GridSize { get; set; } = "3x3";

        [JsonProperty("miniGameType")]
        public string MiniGameType { get; set; } = "";

        [JsonProperty("metadata")]
        public Dictionary<string, object> Metadata { get; set; } = new();

        // Enchanting-specific fields
        [JsonProperty("isEnchantment")]
        public bool IsEnchantment { get; set; }

        [JsonProperty("enchantmentName")]
        public string EnchantmentName { get; set; } = "";

        [JsonProperty("applicableTo")]
        public List<string> ApplicableTo { get; set; } = new();

        [JsonProperty("effect")]
        public Dictionary<string, object> Effect { get; set; } = new();
    }

    /// <summary>
    /// Universal placement data structure for all crafting disciplines.
    /// Supports smithing (grid), refining (hub-and-spoke), alchemy (sequential),
    /// engineering (slot types), and enchanting (pattern-based).
    /// </summary>
    [Serializable]
    public class PlacementData
    {
        [JsonProperty("recipeId")]
        public string RecipeId { get; set; }

        [JsonProperty("discipline")]
        public string Discipline { get; set; }

        // Smithing & Enchanting: Grid-based placement
        [JsonProperty("gridSize")]
        public string GridSize { get; set; } = "";

        [JsonProperty("placementMap")]
        public Dictionary<string, string> PlacementMap { get; set; } = new();

        // Refining: Hub-and-spoke
        [JsonProperty("coreInputs")]
        public List<Dictionary<string, object>> CoreInputs { get; set; } = new();

        [JsonProperty("surroundingInputs")]
        public List<Dictionary<string, object>> SurroundingInputs { get; set; } = new();

        // Alchemy: Sequential
        [JsonProperty("ingredients")]
        public List<Dictionary<string, object>> Ingredients { get; set; } = new();

        // Engineering: Slot types
        [JsonProperty("slots")]
        public List<Dictionary<string, object>> Slots { get; set; } = new();

        // Enchanting: Pattern-based
        [JsonProperty("pattern")]
        public List<Dictionary<string, object>> Pattern { get; set; } = new();

        // Metadata
        [JsonProperty("narrative")]
        public string Narrative { get; set; } = "";

        [JsonProperty("outputId")]
        public string OutputId { get; set; } = "";

        [JsonProperty("stationTier")]
        public int StationTier { get; set; } = 1;
    }
}
